src/mavsdk/core/callback_list_test.cpp

```diff
@@ -98,3 +136,147 @@ TEST(CallbackList, UnsubscribeAllWithNullptr)
+        });
+
+    EXPECT_EQ(callback_count, 0);
+    EXPECT_EQ(nested_callback_count, 0);
+
+    cl(test_value1, test_value2); // Calls 1st callback that adds 2nd callback
+    EXPECT_EQ(callback_count, 1);
+    EXPECT_EQ(nested_callback_count, 0);
+
+    cl(test_value1, test_value2); // Calls 1st callback that adds 3rd callback, calls 2nd callback  
+    EXPECT_EQ(callback_count, 2);
+    EXPECT_EQ(nested_callback_count, 1);
+
+    // Calls 1st callback that adds 4th callback, calls 2nd callback which increments and unsubscribes,
+    // calls the 3rd callback which increments and unsubscribes
+    cl(test_value1, unsub_value);
+    EXPECT_EQ(callback_count, 3);
+    EXPECT_EQ(nested_callback_count, 3);
+
+    // Remove the 1st callback
+    cl.unsubscribe(first_handle);
+
+    // Calls the only remaining 4th callback that increments and unsubscribes
+    cl(test_value1, unsub_value);
+    EXPECT_EQ(callback_count, 3);
+    EXPECT_EQ(nested_callback_count, 4);
+
+    // List is now empty
+    EXPECT_TRUE(cl.empty());
+}
+
+TEST(CallbackList, SubscribeAndUnsubscribeWithinCallbacks2)
+{
+    const int test_value1 = 42;
+    const double test_value2 = 3.14;
+    const int thread_count = 50000;
+    std::atomic<int> callback_count{0};
+    std::atomic<int> nested_callback_count{0};
+    CallbackList<int, double> cl;
+
+    // Define the vector type
+    using HandleType = Handle<int, double>;
+
+    // Create a vector of std::shared_ptr<std::optional<Handle<int, double>>>
+    std::set<HandleType> handle_set;
+    std::mutex mutex_;
+
+    // Lambda function to subscribe a callback as assign the handle to the arg
+    auto addSubCallback = [test_value1, test_value2, &nested_callback_count, &cl](
+                                 std::shared_ptr<std::optional<Handle<int, double>>> handle_ptr) {
+        *handle_ptr = cl.subscribe(
+            [&nested_callback_count, &cl, test_value1, test_value2, handle_ptr](
+                int i, double d) {
+                EXPECT_EQ(i, test_value1);
+                EXPECT_DOUBLE_EQ(d, test_value2);
+                nested_callback_count.fetch_add(1, std::memory_order_relaxed);
+            });
+    };
+
+    // Create multiple threads to simulate concurrent notifications
+    std::vector<std::thread> threads;
+    for (int i = 0; i < thread_count; ++i) {
+        threads.emplace_back([&]() {
+            // Add a burst of subscriptions in a multithreaded env all at once
+            auto handle_ptr = std::make_shared<std::optional<Handle<int, double>>>();
+            addSubCallback(handle_ptr); // this will get deferred sub
+
+            // safely put them in a vec for later comparison
+            std::lock_guard<std::mutex> guard(mutex_);
+            if (handle_ptr && handle_ptr->has_value()) {
+                if (!handle_set.insert(handle_ptr->value()).second) {
+                    EXPECT_TRUE(false);  // handles are not unique
```

Comment:
To make the check fail on main i had to increase the thread count to over 100k and in subscribe() call comment out the check_removals() - not sure why, maybe it was causing some synchronization as a side effect on the rest of the code. In that setting, i could reproduce the failures on main and not on this PR. Another way I checked it was adding a method to pull the _last_id from the callbacklist class, then comparing it with the threadcount, on main (under the conditions mentioned above) it was usually less then the thread_count and not on this PR.

Reply:
Wow! :exploding_head: